pkg/cosign/registryOptions.go

```diff
@@ -0,0 +1,79 @@
+package cosign
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+
+	"github.com/awslabs/amazon-ecr-credential-helper/ecr-login"
+	"github.com/chrismellard/docker-credential-acr-env/pkg/credhelper"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/authn/github"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/google"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	alibabaacr "github.com/mozillazg/docker-credential-acr-helper/pkg/credhelper"
+	ociremote "github.com/sigstore/cosign/v2/pkg/oci/remote"
+	"sigs.k8s.io/release-utils/version"
+)
+
+var uaString = fmt.Sprintf("cosign/%s (%s; %s)", version.GetVersionInfo().GitVersion, runtime.GOOS, runtime.GOARCH)
+
+func UserAgent() string {
+	return uaString
+}
+
+type RegistryOptions struct {
+	AllowInsecure      bool
+	AllowHTTPRegistry  bool
+	KubernetesKeychain bool
+	TagPrefix          string
+	Keychain           authn.Keychain
+}
+
+func (o *RegistryOptions) ClientOpts(ctx context.Context) ([]ociremote.Option, error) {
+	opts := []ociremote.Option{ociremote.WithRemoteOptions(o.GetRegistryClientOpts(ctx)...)}
+	if o.TagPrefix != "" {
+		opts = append(opts, ociremote.WithPrefix(o.TagPrefix))
+	}
+	targetRepoOverride, err := ociremote.GetEnvTargetRepository()
+	if err != nil {
+		return nil, err
+	}
+	if (targetRepoOverride != name.Repository{}) {
+		opts = append(opts, ociremote.WithTargetRepository(targetRepoOverride))
+	}
+	return opts, nil
+}
+
+func (o *RegistryOptions) GetRegistryClientOpts(ctx context.Context) []remote.Option {
```

Comment:
aren't we using the options from the registry client ?

Reply:
I will update it to the use `GCRRemoteOpts` once Amit's PR is merged: https://github.com/kyverno/kyverno/pull/7790

Reply:
Is there no other option ?

Reply:
Should I remove this function after https://github.com/kyverno/kyverno/pull/7806 is merged? As this keychain will be overwritten anyways by `BuildRemoteOption`

Reply:
The idea is that we use the same keychain as the registry client.

Reply:
@eddycharly I removed the registryOptions as client.go will overwrite the registry options anyways, and I added the missing parts to `client.go` in this commit: [`7d675bc` (#7248)](https://github.com/kyverno/kyverno/pull/7248/commits/7d675bce35c1a5c97a76a1bbe0d7b22923c486f1)